### ISSUE TITLE
[Bug] GPU memory usage increase when use backend model inference

### ISSUE THREAD

**Issue by jewelc92** [open-mmlab/mmdeploy#2220]
### Checklist

- [X] I have searched related issues but cannot get the expected help.
- [X] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/tree/main/docs/en/faq.md) but cannot get the expected help.
- [X] 3. The bug has not been fixed in the latest version.

### Describe the bug

I use mmpretrain to train a ResNet-50 model, and use mmdeploy convert pth to onnx and tensorRT, this is the result comparision
![image](https://github.com/open-mmlab/mmdeploy/assets/29191441/6293ed7c-4e75-44ce-8937-db68e912ad35)
Why when use backend model inference, the memory usage increase so much? but when use sdk inference, the memory usage decrease?

The other question is when use backend inference ,the result is complete, it has the pred_score for each class, so I can use my own postprocess. But when use sdk inference, the result only has the pred label and the corresponding pred_score, it doesn't have pred_score for other classes. My question is when use backend infernce, how to reduce memory usage? And when use sdk inference, how to output complete pred result?

### Reproduction

Following is my convert model command line
**convert onnx**
`python tools/deploy.py 
configs/mmpretrain/classification_onnxruntime_static.py
/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_tw_junc_box.py
/disk1/tmp/OpenMMLab/mmpretrain/work_dirs/resnet50_tw_junc_box/epoch_100.pth
/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg 
--work-dir /disk1/tmp/mmdeploy_save_model/mmcls/onnx_static 
--device cuda:0 --dump-info`
**convert tensorrt**
`python tools/deploy.py 
configs/mmpretrain/classification_tensorrt_static-224x224.py
/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_tw_junc_box.py
/disk1/tmp/OpenMMLab/mmpretrain/work_dirs/resnet50_tw_junc_box/epoch_100.pth
/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg 
--work-dir /disk1/tmp/mmdeploy_save_model/mmcls/tensorrt_static_224x224
--device cuda:0 --dump-info`
**convert sdk**
`python tools/deploy.py 
configs/mmpretrain/classification_sdk_dynamic.py
/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_tw_junc_box.py
/disk1/tmp/OpenMMLab/mmpretrain/work_dirs/resnet50_tw_junc_box/epoch_100.pth
/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg 
--work-dir /disk1/tmpmmdeploy_save_model/mmcls/sdk_dynamic
--device cuda:0 --dump-info`

### And the following is my inference code
**pytorch inference**
```
from tqdm import tqdm
from mmpretrain import ImageClassificationInferencer

inferencer = ImageClassificationInferencer(
   model='/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_20230627.py',
   pretrained='/disk1/tmp/OpenMMLab/mmpretrain/work_dirs/resnet50_20230627/epoch_100.pth',
   device='cuda')
result = inferencer(
   ['/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg'],
   show_dir="/disk1/tmp/result/")
```

**onnxruntime inference**
```
from mmdeploy.apis.utils import build_task_processor
from mmdeploy.utils import get_input_shape, load_config
import torch
from tqdm import tqdm

deploy_cfg = '/disk1/tmp/OpenMMLab/mmdeploy/configs/mmpretrain/classification_onnxruntime_static.py'
model_cfg = '/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_20230627.py'
device = 'cuda:0'
backend_model = ['/disk1/tmp/mmdeploy_save_model/mmcls/onnx_static/end2end.onnx']
image = '/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg'

# read deploy_cfg and model_cfg
deploy_cfg, model_cfg = load_config(deploy_cfg, model_cfg)

# build task and backend model
task_processor = build_task_processor(model_cfg, deploy_cfg, device)
model = task_processor.build_backend_model(backend_model)

# process input image
input_shape = get_input_shape(deploy_cfg)
model_inputs, _ = task_processor.create_input(image, input_shape)

# do model inference
with torch.no_grad():
    result = model.test_step(model_inputs)
print(result)
```

**tensorrt inference** 
```
from mmdeploy.apis.utils import build_task_processor
from mmdeploy.utils import get_input_shape, load_config
import torch
from tqdm import tqdm

deploy_cfg = '/disk1/tmp/OpenMMLab/mmdeploy/configs/mmpretrain/classification_tensorrt_static-224x224.py'
model_cfg = '/disk1/tmp/OpenMMLab/mmpretrain/configs/resnet/resnet50_20230627.py'
device = 'cuda:0'
backend_model = ['/disk1/tmp/mmdeploy_save_model/mmcls/tensorrt_static_224x224/end2end.engine']
image = '/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg'

# read deploy_cfg and model_cfg
deploy_cfg, model_cfg = load_config(deploy_cfg, model_cfg)

# build task and backend model
task_processor = build_task_processor(model_cfg, deploy_cfg, device)
model = task_processor.build_backend_model(backend_model)

# process input image
input_shape = get_input_shape(deploy_cfg)
model_inputs, _ = task_processor.create_input(image, input_shape)

# do model inference
with torch.no_grad():
    result = model.test_step(model_inputs)
print(result)
```

**sdk inference**
```
from mmdeploy_runtime import Classifier
import cv2
from tqdm import tqdm

img_path = '/disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg'
img = cv2.imread(img_path)

# create a classifier
classifier = Classifier(model_path='/disk1/tmp/mmdeploy_save_model/mmcls/tensorrt_static_224x224/', device_name='cuda', device_id=0)
# perform inference
result = classifier(img)
# show inference result
print(result)
for label_id, score in result:
    print(label_id, score)
```

### Following is the the output of print(result)
**pytorch**
```
[{'pred_scores': array([0.00433247, 0.00552603, 0.9901416 ], dtype=float32), 
  'pred_label': 2, 'pred_score': 0.9901415705680847, 'pred_class': 'no_hegai'},
 {'pred_scores': array([0.00433247, 0.00552603, 0.9901416 ], dtype=float32), 
  'pred_label': 2, 'pred_score': 0.9901415705680847, 'pred_class': 'no_hegai'}]
```
**onnxruntime & tensorrt**
```
[<DataSample(

META INFORMATION
    ori_shape: (585, 913)
    scale_factor: (0.24534501642935377, 0.38290598290598293)
    img_shape: (224, 224)
    num_classes: 3
    img_path: /disk1/tmp/model_data/cls/2023.02.01/mmcls/val/2/VI_BACK_20221221182654_TP2_2.jpg

DATA FIELDS
    pred_score: tensor([0.0043, 0.0055, 0.9902], device='cuda:0')
    pred_label: tensor([2], device='cuda:0')

) at 0x7f32cb13d190>]
```
**sdk**
```
[(2, 0.9896659255027771)]
```
### Environment

```Shell
06/28 04:00:50 - mmengine - INFO - 

06/28 04:00:50 - mmengine - INFO - **********Environmental information**********
06/28 04:00:51 - mmengine - INFO - sys.platform: linux
06/28 04:00:51 - mmengine - INFO - Python: 3.7.7 (default, May  7 2020, 21:25:33) [GCC 7.3.0]
06/28 04:00:51 - mmengine - INFO - CUDA available: True
06/28 04:00:51 - mmengine - INFO - numpy_random_seed: 2147483648
06/28 04:00:51 - mmengine - INFO - GPU 0,1,2,3: NVIDIA GeForce RTX 3090
06/28 04:00:51 - mmengine - INFO - CUDA_HOME: /usr/local/cuda
06/28 04:00:51 - mmengine - INFO - NVCC: Cuda compilation tools, release 11.3, V11.3.58
06/28 04:00:51 - mmengine - INFO - GCC: gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
06/28 04:00:51 - mmengine - INFO - PyTorch: 1.10.2+cu113
06/28 04:00:51 - mmengine - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.2, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

06/28 04:00:51 - mmengine - INFO - TorchVision: 0.11.3+cu113
06/28 04:00:51 - mmengine - INFO - OpenCV: 4.7.0
06/28 04:00:51 - mmengine - INFO - MMEngine: 0.7.4
06/28 04:00:51 - mmengine - INFO - MMCV: 2.0.0
06/28 04:00:51 - mmengine - INFO - MMCV Compiler: GCC 9.3
06/28 04:00:51 - mmengine - INFO - MMCV CUDA Compiler: 11.3
06/28 04:00:51 - mmengine - INFO - MMDeploy: 1.1.0+90f17f2
06/28 04:00:51 - mmengine - INFO - 

06/28 04:00:51 - mmengine - INFO - **********Backend information**********
06/28 04:00:51 - mmengine - INFO - tensorrt:    8.2.5.1
06/28 04:00:51 - mmengine - INFO - tensorrt custom ops: Available
06/28 04:00:51 - mmengine - INFO - ONNXRuntime: None
06/28 04:00:51 - mmengine - INFO - ONNXRuntime-gpu:     1.8.1
06/28 04:00:51 - mmengine - INFO - ONNXRuntime custom ops:      Available
06/28 04:00:51 - mmengine - INFO - pplnn:       None
06/28 04:00:51 - mmengine - INFO - ncnn:        None
06/28 04:00:51 - mmengine - INFO - snpe:        None
06/28 04:00:51 - mmengine - INFO - openvino:    2022.3.0
06/28 04:00:51 - mmengine - INFO - torchscript: 1.10.2+cu113
06/28 04:00:51 - mmengine - INFO - torchscript custom ops:      NotAvailable
06/28 04:00:51 - mmengine - INFO - rknn-toolkit:        None
06/28 04:00:51 - mmengine - INFO - rknn-toolkit2:       None
06/28 04:00:51 - mmengine - INFO - ascend:      None
06/28 04:00:51 - mmengine - INFO - coreml:      None
06/28 04:00:51 - mmengine - INFO - tvm: None
06/28 04:00:51 - mmengine - INFO - vacc:        None
06/28 04:00:51 - mmengine - INFO - 

06/28 04:00:51 - mmengine - INFO - **********Codebase information**********
06/28 04:00:52 - mmengine - INFO - mmdet:       3.0.0
06/28 04:00:52 - mmengine - INFO - mmseg:       1.0.0
06/28 04:00:52 - mmengine - INFO - mmpretrain:  1.0.0rc8
06/28 04:00:52 - mmengine - INFO - mmocr:       None
06/28 04:00:52 - mmengine - INFO - mmagic:      None
06/28 04:00:52 - mmengine - INFO - mmdet3d:     None
06/28 04:00:52 - mmengine - INFO - mmpose:      None
06/28 04:00:52 - mmengine - INFO - mmrotate:    None
06/28 04:00:52 - mmengine - INFO - mmaction:    None
06/28 04:00:52 - mmengine - INFO - mmrazor:     None
```

**Comment by irexyc** [comment]
@jewelc92 

When you use backend inference api, the inner process will import torch and use some torch apis, so the memory will be larger than using pure pytorch or pure backend(ort, trt, etc.). The SDK only use backend library so the memory will be less than backend inference api.

>  But when use sdk inference, the result only has the pred label and the corresponding pred_score, it doesn't have pred_score for other classes.

I guess you want all class output? You can edit the `pipeline.json`, change `topk` to the number of model num_classes output should be work.


```
{
    "type": "Task",
    "module": "mmcls",
    "name": "postprocess",
    "component": "LinearClsHead",
    "params": {
        "num_classes": 1000,
        "in_channels": 512,
        "loss": {
            "type": "CrossEntropyLoss",
            "loss_weight": 1.0
        },
        "topk": 5
    },
    "output": [
        "post_output"
    ],
    "input": [
        "infer_output"
    ]
}

**Comment by jewelc92** [comment]
Thanks, my probelm solved
